blocks/library/navigation-menu/block.js

```diff
@@ -0,0 +1,216 @@
+}
+
+class NavigationMenuBlock extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.setMenu = this.setMenu.bind( this );
+		this.renderMenu = this.renderMenu.bind( this );
+		this.renderMenuLevel = this.renderMenuLevel.bind( this );
+	}
+
+	setMenu( value ) {
+		const { setAttributes } = this.props;
+		if ( value ) {
+			setAttributes( { selected: value } );
+		}
+	}
+
+	renderMenuLevel( items ) {
+		if ( ! items || items.length < 1 ) {
+			return null;
+		}
+
+		const { layout } = this.props.attributes;
+		const isTopLevel = items[ 0 ].menu_item_parent === 0;
+		const isHorizontal = layout === 'horizontal';
+		const className = classnames( {
+			[ this.props.className ]: isTopLevel,
+			'is-horizontal': isHorizontal,
+			'is-vertical': ! isHorizontal,
+		} );
+
+		return (
+			<ul
+				key="navigation-menu"
+				className={ className }
+			>
+				{ items.map( ( item, i ) => {
+					return (
+						<MenuItem key={ i } item={ item }>
+							{ item.children && this.renderMenuLevel( item.children ) }
+						</MenuItem>
+					);
+				} ) }
+			</ul>
+		);
+	}
+
+	renderMenu() {
+		const { data, isLoading } = this.props.items;
+		const customizerUrl = '';
+
+		if ( isUndefined( data ) || isEmpty( data ) || isLoading ) {
+			return (
+				<Placeholder
+					key="navigation-menu"
+					icon="menu"
+					label={ __( 'Navigation Menu' ) }
+				>
+					{ ! Array.isArray( data ) ?
+						<Spinner /> :
+						<Button href={ `${ customizerUrl }?autofocus%5Bpanel%5D=nav_menus` }>
+							{ __( 'No items found in this menu.' ) }
+						</Button>
+					}
+				</Placeholder>
+			);
+		}
+
+		const nestedData = getMenuWithChildren( data );
+		return this.renderMenuLevel( nestedData );
+	}
+
+	render() {
+		const { attributes, focus, setAttributes } = this.props;
+		const { align, layout, selected } = attributes;
+		const { data: menus, isLoading } = this.props.menus;
+
+		const menuSelectControl = (
+			<SelectControl
+				label={ __( 'Select an existing menu' ) }
+				value={ selected }
+				onChange={ this.setMenu }
+				options={ getOptionsFromMenu( menus, selected ) }
+			/>
+		);
+
+		const inspectorControls = focus && (
+			<InspectorControls key="inspector">
+				<h3>{ __( 'Menu Settings' ) }</h3>
+				{ ! isLoading && menuSelectControl }
+			</InspectorControls>
+		);
+
+		const layoutControls = [
+			{
+				icon: 'list-view',
+				title: __( 'Vertical View' ),
+				onClick: () => setAttributes( { layout: 'vertical' } ),
+				isActive: layout === 'vertical',
+			},
+			{
+				icon: 'grid-view',
+				title: __( 'Horizontal View' ),
+				onClick: () => setAttributes( { layout: 'horizontal' } ),
+				isActive: layout === 'horizontal',
+			},
+		];
+
+		let displayBlock = (
+			<MenuPlaceholder
+				key="navigation-menu"
+				menus={ menus }
+				selected={ false }
+				setMenu={ this.setMenu } >
+				{ menuSelectControl }
+			</MenuPlaceholder>
+		);
+
+		if ( !! selected ) {
+			displayBlock = this.renderMenu();
```

Comment:
We could avoid having to implement `renderMenu` here which is a duplication of `wp_nav_menu()` if we used `ServerSideRender` proposed in #780.